### PR TITLE
dcos-1733 validate app/group updates

### PR DESCRIPTION
**Pull request by tamarrow**

### CODE REVIEW

**Comment by tamarrow** [comment]
@jsancio bump

File: cli/dcoscli/marathon/main.py [hunk 1]
```diff
@@ -596,15 +599,61 @@ def _update(app_id, properties, force):
...
-    app_resource = _parse_properties(properties, _app_schema())
     deployment = client.update_app(app_id, app_resource, force)
 
     emitter.publish('Created deployment {}'.format(deployment))
     return 0
 
 
+def _validate_update(current_resource, properties, schema):
+    """
+    Validate resource ("app" or "group") update
```
Comment (jsancio) [inline cli/dcoscli/marathon/main.py]: Fix the the parameter descriptions below. The syntax doesn't look correct. E.g. `:param current_resource: ...` and `:type current_resource: ...`

File: cli/dcoscli/marathon/main.py [hunk 2]
```diff
@@ -596,15 +599,61 @@ def _update(app_id, properties, force):
...
+
+    return 0
+
+
+def _clean_up_properties(properties):
+    """
+    Remove task properties and nulls from app definition
+
+    :param properties: resource JSON
+    :type: dict
```
Comment (jsancio) [inline cli/dcoscli/marathon/main.py]: missing parameter name.

File: cli/dcoscli/marathon/main.py [hunk 3]
```diff
@@ -596,15 +599,61 @@ def _update(app_id, properties, force):
...
+    :rtype: int
+    """
+    updated_resource = _clean_up_properties(current_resource.copy())
+    updated_resource.update(properties)
+
+    errs = util.validate_json(updated_resource, schema)
+    if errs:
+        raise DCOSException(util.list_to_err(errs))
+
+    return 0
```
Comment (jsancio) [inline cli/dcoscli/marathon/main.py]: Why are we returning zero? We never use the value at the call site.

**Comment by jsancio** [comment]
Let's add some tests or are we happy with the existing test cases? What about the negative case when the update fails to validate. Are we testing that?

File: cli/dcoscli/marathon/main.py (hunk 1 above)
Comment (tamarrow) [inline cli/dcoscli/marathon/main.py]: done

File: cli/dcoscli/marathon/main.py (hunk 3 above)
Comment (tamarrow) [inline cli/dcoscli/marathon/main.py]: removed

File: cli/dcoscli/marathon/main.py (hunk 2 above)
Comment (tamarrow) [inline cli/dcoscli/marathon/main.py]: done

**Comment by tamarrow** [comment]
I'm happy with the tests (they found the initial problem!) We have `test_update_missing_field`, `test_update_bad_type`, and `test_update_app`

File: cli/dcoscli/marathon/main.py [hunk 4]
```diff
@@ -596,15 +599,58 @@ def _update(app_id, properties, force):
...
+    """
+    updated_resource = _clean_up_properties(current_resource.copy())
+    updated_resource.update(properties)
+
+    errs = util.validate_json(updated_resource, schema)
+    if errs:
+        raise DCOSException(util.list_to_err(errs))
+
+
+def _clean_up_properties(properties):
```
Comment (jsancio) [inline cli/dcoscli/marathon/main.py]: Let's change this method name to something more descriptive. Maybe `_clean_up_app_definition`? Is app definition the correct term? The schema has apps and groups, no? Does Marathon have a name for that schema?

**Comment by jsancio** [comment]
Looks good to me! Thanks! One small comment left.

File: cli/dcoscli/marathon/main.py (hunk 4 above)
Comment (tamarrow) [inline cli/dcoscli/marathon/main.py]: I've been using `resource` for both. maybe `_clean_up_resource_definition`?
